faiss/impl/HNSW.cpp

```diff
@@ -966,13 +1219,28 @@ HNSWStats HNSW::search(
     }
 
     int ef = std::max(efSearch, k);
-    if (bounded_queue) { // this is the most common branch
+    if (bounded_queue) { // this is the most common branch, for now we only
+                         // support Panorama search in this branch
         MinimaxHeap candidates(ef);
 
         candidates.push(nearest, d_nearest);
 
-        search_from_candidates(
-                *this, qdis, res, candidates, vt, stats, 0, 0, params);
+        if (is_panorama) {
```

Comment:
An alternative would be to make `search_from_candidates` a class function, and then create an `HNSWPanorama` class that overrides `search_from_candidates`.

We could then store `cum_sums` and other relevant state within the `HNSWPanorama` object. This is my preferred option, but it would mean that the function is no longer in the `faiss` namespace. I wanted to check whether that has any consequences.